server/aesdsocket.c

```diff
@@ -1,117 +1,424 @@
+    exit(SUCCESS);
+}
 
 /*************************************************************************
  *             Signal Handler Function                                   *
  *************************************************************************/
+ 
 void signalhandler (int signal)
-{
-    // Logs message to the syslog “Caught signal, exiting” when SIGINT or SIGTERM is received.
+{ 
+
     if (signal == SIGINT || signal == SIGTERM )
+    {   
+        signal_exit = true;
+        
+    	shutdown(sockfd, SHUT_RDWR);
+    	cleanup();
+    }   
+}
+
+/*************************************************************************
+ *               Time Handler Function                                   *
+ *************************************************************************/
+ 
+void timehandler(int signal)
+{ 
+    if (signal == SIGALRM)
+    {          
+        timer_exit = true;
+    }   
+}
+
+/*************************************************************************
+ *                  Multithread_handler Function                         *
+ *************************************************************************/
+ 
+void *multithread_handler(void *new_client_node)
+{
+	struct slist_client_s *thread_param = (struct slist_client_s*)new_client_node;
+	
+	/*************************************************************************
+     *                            Receive                                    *
+     *************************************************************************/ 
+    pthread_mutex_lock(&lock);
+    // Receives data over the connection and appends to file /var/tmp/aesdsocketdata, creating this file if it doesn’t exist.
+    if ((file_ptr = fopen(DATA_FILE, "a+")) == NULL) //opens file in append and update mode and checks if error
     {
-        syslog(LOG_INFO, "Caught signal, exiting");
-        close(sockfd);
-        remove(DATA_FILE);
-        closelog();
-        exit(SUCCESS);
+        syslog(LOG_ERR,"Error while opening given file; fopen() failure\n"); //syslog error
+	printf("Error! fopen() failure\n"); //prints error
+	return NULL;
     }
+    char buffer[BUFFER_SIZE];
+    	
+    while(1)
+    {
+        // Ref: [12] man page
+        // receive - returns the number of bytes actually read into the buffer
+        // int recv(int sockfd, void *buf, int len, int flags);
+        int num_bytes = recv(thread_param->newfd, buffer, sizeof(buffer), 0); // Receive data from the newfd socket into the buffer. It reads up to sizeof(buffer) bytes at a time.
+        if (num_bytes == RET_FAILURE)
+        {
+        	syslog(LOG_ERR,"Error while receiving; recv() failure\n"); //syslog error
+			printf("Error! recv() failure/n");                           //prints error	
+			break;
+        }
+        	
+        // Ref: [13] man page
+        // Received data written to file
+        // size_t fwrite(const void *ptr, size_t size, size_t count, FILE *stream);
+        fwrite(buffer, 1, num_bytes, file_ptr);
```

Comment:
what if \n is not the end of the input string? say "Hello \n I'm Swathi" only "Hello " should be written to the file and sent back